File: io-hotmoka-network-thin-client/io-hotmoka-network-thin-client-kt/src/main/kotlin/io/hotmoka/network/thin/client/models/requests/MethodCallTransactionRequestModel.kt
package io.hotmoka.network.thin.client.models.requests

import io.hotmoka.network.thin.client.models.signatures.MethodSignatureModel
import io.hotmoka.network.thin.client.models.values.StorageReferenceModel
import io.hotmoka.network.thin.client.models.values.StorageValueModel
import io.hotmoka.network.thin.client.models.values.TransactionReferenceModel

/**
 * The model of a method call transaction request.
 */
abstract class MethodCallTransactionRequestModel(
        signature: String,
        caller: StorageReferenceModel,
        nonce: String,
        classPath: TransactionReferenceModel,
        chainId: String,
        gasLimit: String,
        gasPrice: String,
        val method: MethodSignatureModel,
        val actuals: List<StorageValueModel>
) : NonInitialTransactionRequestModel(
        signature,
        caller,
        nonce,
        classPath,
        chainId,
        gasLimit,
        gasPrice
)

File: io-hotmoka-network-thin-client/io-hotmoka-network-thin-client-js/src/models/requests/MethodCallTransactionRequestModel.ts
import {MethodSignatureModel} from "../signatures/MethodSignatureModel";
import {StorageValueModel} from "../values/StorageValueModel";
import {StorageReferenceModel} from "../values/StorageReferenceModel";
import {TransactionReferenceModel} from "../values/TransactionReferenceModel";
import {CodeExecutionTransactionRequestModel} from "./CodeExecutionTransactionRequestModel";
import {MarshallingContext} from "../../internal/marshalling/MarshallingContext";
import {HotmokaException} from "../../internal/exception/HotmokaException";

/**
 * The model of a method call transaction request.
 */
export abstract class MethodCallTransactionRequestModel extends CodeExecutionTransactionRequestModel {
    method: MethodSignatureModel

    protected constructor(
        caller: StorageReferenceModel,
        nonce: string,
        classpath: TransactionReferenceModel,
        gasLimit: string,
        gasPrice: string,
        method: MethodSignatureModel,
        actuals: Array<StorageValueModel>) {

        super(caller, nonce, classpath, gasLimit, gasPrice, actuals)

        if (method === null || method === undefined) {
            throw new HotmokaException("method signature cannot be null or undefined")
        }

        const formalsLength = method.formals ? method.formals.length : 0
        const actualsLength = actuals ? actuals.length : 0

        if (formalsLength !== actualsLength) {
            throw new HotmokaException("argument count mismatch between formals and actuals")
        }

        this.method = method
    }

    protected intoWithoutSignature(context: MarshallingContext): void {
        super.intoWithoutSignature(context)
        this.method.into(context)
    }
}